flocker/volume/filesystems/openstack.py

```diff
@@ -0,0 +1,240 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+ZFS APIs.
+"""
+
+from __future__ import absolute_import
+
+import os
+import string
+
+from subprocess import check_call
+import time
+
+from characteristic import with_cmp, with_repr
+
+from zope.interface import implementer
+
+from eliot import Logger
+
+from twisted.python.filepath import FilePath
+from twisted.internet.defer import succeed
+from twisted.application.service import Service
+
+from .interfaces import IFilesystem, IStoragePool
+
+from .._model import VolumeSize, VolumeName
+
+import pyrax
+
+import netifaces
+import ipaddr
+
+
+def get_public_ips():
+    ips = []
+    for interface in netifaces.interfaces():
+        interface_addresses = netifaces.ifaddresses(interface)
+        ipv4addresses = interface_addresses.get(netifaces.AF_INET, [])
+        for address in ipv4addresses:
+            ip = ipaddr.IPv4Address(address['addr'])
+            if not ip.is_private:
+                ips.append(ip)
+    return ips
+
+
+def driver_from_environment():
+    # Notes on how to get this working are in https://clusterhq.atlassian.net/browse/FLOC-1147
+    username = os.environ.get('OPENSTACK_API_USER')
+    api_key = os.environ.get('OPENSTACK_API_KEY')
+
+    ctx = pyrax.create_context(
+        id_type="rackspace", username=username, api_key=api_key)
+    ctx.authenticate()
+    compute = ctx.get_client('compute', 'DFW')
+    volume = ctx.get_client('volume', 'DFW')
+
+    return compute, volume
+
+
+def next_device():
+    """
+    Can't just use the dataset name as the block device name
+    inside the node, nor volume.id nor random_name. You can't
+    even leave it blank; auto is not supported.
```

Comment:
We should discuss this too. In view of @lukemarsden 's comments that servers are limited to 24 rackspace block devices and what we read in:
 * http://www.rackspace.com/knowledge_center/article/cloud-block-storage-overview#limits-of-cbs

> Limits of Cloud Block Storage
>     100 GB to 1 TB / Volume
>    14 Volumes max / Server - Operating system (OS) dependent
>    50 Volumes max / region OR 10 TB max/ region (whichever is first) - This is the default for all new customers. > Customers can request limit increases. Please contact Rackspace Support, your Account Manager, or your Service Delivery Manager for more information.